pkg/client/opensergo_client.go

```diff
@@ -0,0 +1,215 @@
+	return openSergoClient.subscriberRegistry
+}
+
+// RegisterSubscribeInfo
+//
+// registry Subscriber
+// if subscribeConfigClient in OpenSergoClient is initialized then subscribe config-data by subscribeKey
+func (openSergoClient *OpenSergoClient) RegisterSubscribeInfo(subscribeInfo *SubscribeInfo) *OpenSergoClient {
+	subscribers := openSergoClient.GetSubscriberRegistry().GetSubscribersOf(subscribeInfo.subscribeKey)
+	for _, subscriber := range subscribeInfo.subscribers {
+		openSergoClient.GetSubscriberRegistry().RegisterSubscriber(subscribeInfo.subscribeKey, subscriber)
+	}
+
+	if openSergoClient.status.Load().(OpensergoClientStatus) == started && len(subscribers) == 0 {
+		openSergoClient.SubscribeConfig(subscribeInfo.subscribeKey)
+	}
+	return openSergoClient
+}
+
+// KeepAlive
+//
+// keepalive OpenSergoClient
+func (openSergoClient *OpenSergoClient) keepAlive() {
+	status := openSergoClient.status.Load().(OpensergoClientStatus)
+	if status == initial || status == interrupted {
+		logging.Info("[OpenSergo SDK] try to restart openSergoClient...")
+		openSergoClient.Start()
+	}
+	time.Sleep(time.Duration(10) * time.Second)
+	openSergoClient.keepAlive()
+}
+
+// Start
+//
+// to start the OpenSergoClient,
+// and start the opensergoClientObserver for OpenSergoClient
+func (openSergoClient *OpenSergoClient) Start() {
+
+	logging.Info("[OpenSergo SDK] OpenSergoClient is starting...")
+
+	// keepalive OpensergoClient
+	if openSergoClient.status.Load().(OpensergoClientStatus) == initial {
+		logging.Info("[OpenSergo SDK] open keepalive() goroutine to keep OpensergoClient alive")
+		go openSergoClient.keepAlive()
+	}
+
+	openSergoClient.status.Store(starting)
+
+	subscribeConfigStream, err := (*openSergoClient.transportServiceClient).SubscribeConfig(context.Background())
+	if err != nil {
+		logging.Error(err, "[OpenSergo SDK] SubscribeConfigStream can not connect.")
+		openSergoClient.status.Store(interrupted)
+		return
+	}
+	openSergoClient.subscribeConfigStream = subscribeConfigStream
+
+	if openSergoClient.subscribeConfigStreamObserver == nil {
+		openSergoClient.subscribeConfigStreamObserver = NewSubscribeConfigStreamObserver(openSergoClient)
+	}
+	openSergoClient.subscribeConfigStreamObserver.Start()
+
+	logging.Info("[OpenSergo SDK] begin to subscribe config-data...")
+	subscribersInRegistry := openSergoClient.GetSubscriberRegistry().GetSubscribersAll()
+	subscribersInRegistry.Range(func(key, value interface{}) bool {
+		subscribeKey := key.(subscribe.SubscribeKey)
+		openSergoClient.SubscribeConfig(subscribeKey)
+		return true
+	})
+
+	logging.Info("[OpenSergo SDK] OpenSergoClient is started")
+	openSergoClient.status.Store(started)
+}
+
+// SubscribeConfig
+//
+// send a subscribe request to opensergo-control-plane,
+// and return the result of subscribe config-data from opensergo-control-plane.
+func (openSergoClient OpenSergoClient) SubscribeConfig(subscribeKey subscribe.SubscribeKey) bool {
```

Comment:
Please use pointer receiver here, and please re-check other part of code :)

Reply:
has fixed. and also has checked all the receivers of function.